text/0000-deref-move.md

```diff
@@ -0,0 +1,325 @@
+not seem worthwhile to move further with it at this time.
+
+## Weakening `DerefMut` bound
+
+The `DerefMut` bound on `DerefMove` and, in particular, the `Deref` bound that
+it implies, constrain implementors somewhat. It requires that `Deref` be
+implementable, so the implementation is not allowed to overload dereferencing in
+such a way that it produces a new value each time. This would not be useful for
+non-`Copy` types, since they would be consumed, but a `Copy` type could in this
+fashion allow itself to be copied and then produce a new value on each
+derefence. This would be useful to avoid redefining methods on the outer type.
+
+This very much goes against the intent of the `Deref` traits to be restricted
+smart pointers, however. `Deref` is currently the only way for methods of one
+type to be callable on another, but it is deliberately limited in scope. For
+instance, one might be tempted to implement `Deref` on a newtype for this
+purpose, but this is considered unidiomatic. Additionally, this is a bandaid for
+the fact that newtypes would also often like to be able to inherent trait
+implementations, and inhereting methods is not sufficient for that.
+
+The additional requirement is to avoid complication of the following case:
+
+```rust
+struct RefMethod();
+
+trait Consume : Sized {
+    fn consume(self) {}
+}
+
+impl<'a> Consume for &'a mut RefMethod {}
+
+struct BadDeref<'a> (&'a mut RefMethod);
+
+impl<'a> Deref for BadDeref<'a> {
+    type Target = RefMethod;
+    fn deref(&self) -> &Self::Target { &self.0 }
+}
+
+impl<'a> DerefMove for BadDeref<'a> {
+    fn deref_move(self) -> Self::Target { self.0 }
+}
+
+fn main() {
+    let mut r = RefMethod{};
+    let b = BadDeref(&mut r);
+    b.consume();
+    b.consume();
+}
+```
+
+In this case, the first `b.consume()` is legal, but the second `b.consume()` is
+not since `b` was already moved out of. But if `BadDeref` also implements
+`DerefMut`, then the first `b.consume()` becomes illegal because `deref_mut`
+becomes preferred and `b` is immutable. By requiring that `DerefMove :
+DerefMut`, this edge case cannot be encountered: this code will always be
+illegal unless `b` is made mutable.
+
+# Prior art
+[prior-art]: #prior-art
+
+The compiler special case for `Box` is the most compelling prior art. This RFC
+is written with that behaviour in mind; one of the most important criteria here
+is that `Box` should remain completely backwards-compatible. In case of any
+conflict between current `Box` behaviour and the behaviour specified in this
+RFC that isn't explicitly called out as being different, it should be considered
+a bug in the RFC and either this RFC should be corrected or more discussion
+should ensue.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+## Implementors
+
+It's not clear to me which, if any, of the smart pointers above should implement
+`DerefMove` other than `Box`. It requires discussion of whether inadvertently
```

Comment:
`Vec` should likely have an implementation; I expect it can't right now because of sizedness issues.  Actually, that suggests to me that allowing by-value DSTs is pretty important to making `DerefMove` work well.

Reply:
That's a good point. I will revise the RFC to discuss the interaction with by-value DSTs more, since that seems very fundamental.